### PR TITLE
Feat: Decode error messages from Looker API

### PR DESCRIPTION
**Pull request by jessicasomaiya**
## Description

This PR is to decode the error messages coming back from the API, as the Looker API documentation defines the shape of the errors. For status code `422`: `http.StatusUnprocessableEntity` the response is a `ValidationError`. For other non-2xx status codes the error is of type `Error`. 

This PR also addresses the issue of identifying when the error response has status code `404`. Not found errors are wrapped in a well known error type so they can be identified and these cases can be specifically handled. 

## How does this implementation work? 

As the error models are defined in the `v4` package, the `rtl` package cannot import the models as this would lead to a cyclic dependancy. This means the `Do` method https://github.com/looker-open-source/sdk-codegen/blob/main/go/rtl/auth.go#L89 must expose the error status and the body so the body can be decoded in the v4  package.

This means rather than the `Do` method decoding the error response on behalf of each method, each method must decode the response. To do this, I've changed the generator to wrap every error in a `deserialiseError`.

### CODE REVIEW

File: go/rtl/auth.go [hunk 1]
```diff
@@ -119,7 +120,10 @@ func (s *AuthSession) Do(result interface{}, method, ver, path string, reqPars m
 			return fmt.Errorf("response error. status=%s. error parsing error body", res.Status)
 		}
 
-		return fmt.Errorf("response error. status=%s. error=%s", res.Status, string(b))
+		return ResponseError{
```
Comment (mattoddie) [inline go/rtl/auth.go:122]: Can we pass in `res.Status` too as a separate field? Then in the `Error()` method we can reuse it so the error message doesn't technically change (incase anyone was parsing it)

File: go/sdk/v4/errors.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	var errSlice []string
+	for _, m := range *e.Errors {
+		if m.Field != nil && m.Message != nil {
+			errSlice = append(errSlice, fmt.Sprintf("'%s' (%s)", *m.Field, *m.Message))
+		}
+	}
+	return fmt.Sprintf("validation error on fields: %s", strings.Join(errSlice, ", "))
+}
+
+// deseraliseError decodeds the error message depending on the error response status.
```
Comment (mattoddie) [inline go/sdk/v4/errors.go]:
```suggestion
// deseraliseError decodes the error message depending on the error response status.
```

File: go/sdk/v4/errors.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+	re, ok := err.(rtl.ResponseError)
+	if !ok {
+		return err
+	}
+
+	switch re.StatusCode {
+	case http.StatusUnprocessableEntity:
+		e := new(ValidationError)
+		if err := json.Unmarshal(re.Body, &e); err != nil {
```
Comment (mattoddie) [inline go/sdk/v4/errors.go]: `e` is already a pointer so no need to do it again

```suggestion
		if err := json.Unmarshal(re.Body, e); err != nil {
```

File: go/sdk/v4/errors.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+			break
+		}
+		return e
+	case http.StatusNotFound:
+		e := new(Error)
+		if err := json.Unmarshal(re.Body, &e); err != nil {
```
Comment (mattoddie) [inline go/sdk/v4/errors.go]:
```suggestion
		if err := json.Unmarshal(re.Body, e); err != nil {
```

File: go/sdk/v4/errors.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+			break
+		}
+		return errors.Wrap(ErrNotFound, e.Message)
+	default:
+		e := new(Error)
+		if err := json.Unmarshal(re.Body, &e); err != nil {
```
Comment (mattoddie) [inline go/sdk/v4/errors.go]:
```suggestion
		if err := json.Unmarshal(re.Body, e); err != nil {
```

File: go/sdk/v4/errors.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+			break
+		}
+		return e
+	}
+
+	return fmt.Errorf("error unmarshalling body with status: %d, body:%s, error:%s", re.StatusCode, re.Body, err.Error())
```
Comment (mattoddie) [inline go/sdk/v4/errors.go]:
```suggestion
	return fmt.Errorf("error unmarshalling body with status:%d, body:%s, error:%s", re.StatusCode, re.Body, err.Error())
```

File: go/sdk/v4/errors.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	return fmt.Sprintf("validation error on fields: %s", strings.Join(errSlice, ", "))
+}
+
+// deseraliseError decodeds the error message depending on the error response status.
+func deserialiseError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	re, ok := err.(rtl.ResponseError)
```
Comment (mattoddie) [inline go/sdk/v4/errors.go:44]: it might be nicer to use `errors.As` here? https://pkg.go.dev/errors#As

File: go/rtl/auth.go [hunk 8]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"context"
 	"crypto/tls"
+
```
Comment (mattoddie) [inline go/rtl/auth.go]:
```suggestion
```

**Review by mattoddie — CHANGES_REQUESTED** [review]

File: go/sdk/v4/errors.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
+package v4
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/looker-open-source/sdk-codegen/go/rtl"
```
Comment (mattoddie) [inline go/sdk/v4/errors.go:10]:
```suggestion
	"github.com/resolutionlife/looker-sdk-codegen/go/rtl"
```

File: go/rtl/auth.go (hunk 1 above)
Comment (jessicasomaiya) [reply go/rtl/auth.go:122]: Yep I was considering this. I'll make the change 👍🏾

File: go/sdk/v4/errors.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/pkg/errors"
```
Comment (mattoddie) [inline go/sdk/v4/errors.go]: Can't see any need for using this - lets just use inbuilt `errors`

**Review by mattoddie — APPROVED** [review]

**Pull request merged**